### PR TITLE
Add new portworx schedops API to check portworx pod status

### PR DESCRIPTION
**Pull request by Rohit-PX**
Logs with debug statements for Pod Name and Pod status.

http://70.0.0.41:8080/view/torpedo/view/all/job/torpedo-dev/67/console

### CODE REVIEW

**Comment by piyush-nimbalkar** [comment]
why is this check needed? where are we planning to use it?

**Comment by Rohit-PX** [comment]
@piyush-nimbalkar  we do not want to rely on libopenstorage APIs to get the status of the volume drivers. Hence we added this check in WaitDriverUpOnNode to check the status of the PX pods.

**Comment by piyush-nimbalkar** [comment]
Talked to Rohit offline, why this is needed.
@harsh-px If the OSD status is not updated frequently, then apart from checking the OCI Mon container, we should also check if Portworx service is running or not to be completely sure.

File: drivers/volume/portworx/portworx.go [hunk 1]
```diff
@@ -624,6 +624,12 @@ func (d *portworx) WaitDriverUpOnNode(n node.Node) error {
 		return err
 	}
 
+	//Check if PX pod is up
+	if d.schedOps.IsPXPodRunningOnNode(n) != true {
+		return fmt.Errorf("PX pod is not up on node: %s", n.Name)
+	}
+	fmt.Printf("\nPX pod is up on node: %s", n.Name)
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/portworx.go]: use logrus.Infof

File: drivers/volume/portworx/portworx.go [hunk 2]
```diff
@@ -624,6 +624,12 @@ func (d *portworx) WaitDriverUpOnNode(n node.Node) error {
 		return err
 	}
 
+	//Check if PX pod is up
+	if d.schedOps.IsPXPodRunningOnNode(n) != true {
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/portworx.go]: nit: use ! instead of comparing with true

File: drivers/volume/portworx/schedops/schedops.go [hunk 3]
```diff
@@ -29,6 +29,8 @@ type Driver interface {
 	GetServiceEndpoint() (string, error)
 	// UpgradePortworx upgrades portworx to the given docker image and tag
 	UpgradePortworx(image, tag string) error
+	//IsPXPodRunningOnNode returns true if PX pod is up on that node, else returns false
+	IsPXPodRunningOnNode(n node.Node) bool
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/schedops.go]: nit: let's not have "Pod" in this method name as it is a scheduler specific name.
Up to you, we can change this later as well when we support other schedulers.

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 4]
```diff
@@ -451,6 +451,18 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
 	return nil
 }
 
+//Method to validate if Portworx pod is up and running
+func (k *k8sSchedOps) IsPXPodRunningOnNode(n node.Node) bool {
+	pxPods, _ := k8s.Instance().GetPodsByNode(n.Name, "kube-system")
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: fail or return false on error

Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: use PXNamespace const for kube-system

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 5]
```diff
@@ -451,6 +451,18 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
...
+	for _, pod := range pxPods.Items {
+		if strings.Contains(pod.Name, "portworx") && pod.Status.Phase != "Running" {
+			fmt.Printf("\nError on %s Pod: %v is not up yet. Pod Status: %v\n", pod.Status.PodIP, pod.Name, pod.Status.Phase)
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: use logrus

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 6]
```diff
@@ -451,6 +451,18 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
 	return nil
 }
 
+//Method to validate if Portworx pod is up and running
+func (k *k8sSchedOps) IsPXPodRunningOnNode(n node.Node) bool {
+	pxPods, _ := k8s.Instance().GetPodsByNode(n.Name, "kube-system")
+	for _, pod := range pxPods.Items {
+		if strings.Contains(pod.Name, "portworx") && pod.Status.Phase != "Running" {
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: nit: we can use strings.HasPrefix to be a little more strict

File: drivers/volume/portworx/portworx.go [hunk 7]
```diff
@@ -624,6 +624,12 @@ func (d *portworx) WaitDriverUpOnNode(n node.Node) error {
 		return err
 	}
 
+	//Check if PX pod is up
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/portworx.go]: remove the debug line on 609

File: drivers/volume/portworx/schedops/schedops.go (hunk 3 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/schedops.go]: Changed to "IsPXAppRunningOnNode".

File: drivers/volume/portworx/schedops/k8s-schedops.go (hunk 5 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Fixed.

File: drivers/volume/portworx/schedops/k8s-schedops.go (hunk 4 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Fixed.

File: drivers/volume/portworx/schedops/k8s-schedops.go (hunk 6 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Fixed.

File: drivers/volume/portworx/schedops/k8s-schedops.go (hunk 4 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Returning false after printing error.

File: drivers/volume/portworx/portworx.go (hunk 1 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/portworx.go]: Fixed.

File: drivers/volume/portworx/portworx.go (hunk 2 above)
Comment (Rohit-PX) [reply drivers/volume/portworx/portworx.go]: Fixed.

**Comment by harsh-px** [comment]
@piyush-nimbalkar OSD API does update right way. The need for this PR is to check if oci-monitor itself is having issues and not showing ready regardless of what the OSD status is.

File: drivers/volume/portworx/portworx.go [hunk 8]
```diff
@@ -624,6 +623,12 @@ func (d *portworx) WaitDriverUpOnNode(n node.Node) error {
 		return err
 	}
 
+	// Check if PX pod is up
+	if !d.schedOps.IsPXAppRunningOnNode(n) {
+		return fmt.Errorf("PX pod is not up on node: %s", n.Name)
+	}
+	logrus.Infof("\nPX pod is up on node: %s", n.Name)
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/portworx.go]: nit: remove the \n

File: drivers/volume/portworx/schedops/dcos-schedops.go [hunk 9]
```diff
@@ -56,6 +56,11 @@ func (d *dcosSchedOps) UpgradePortworx(image, tag string) error {
 	return nil
 }
 
+func (d *dcosSchedOps) IsPXAppRunningOnNode(n node.Node) bool {
+	// TOOD: Implement this method
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/dcos-schedops.go]: nit: *TODO

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 10]
```diff
@@ -451,6 +451,22 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
...
+//Method to validate if Portworx pod is up and running
+func (k *k8sSchedOps) IsPXAppRunningOnNode(n node.Node) bool {
+	pxPods, err := k8s.Instance().GetPodsByNode(n.Name, PXNamespace)
+	if err != nil {
+		logrus.Errorf("Failed to get apps on node %s", n.Name)
+		return false
+	}
+	for _, pod := range pxPods.Items {
+		if strings.HasPrefix(pod.Name, "portworx") && pod.Status.Phase != "Running" {
+			logrus.Errorf("\nError on %s Pod: %v is not up yet. Pod Status: %v\n", pod.Status.PodIP, pod.Name, pod.Status.Phase)
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: nit: no need for \n

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 11]
```diff
@@ -451,6 +451,22 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
 	return nil
 }
 
+//Method to validate if Portworx pod is up and running
```
Comment (piyush-nimbalkar) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: nit: space after // everywhere

File: drivers/volume/portworx/schedops/k8s-schedops.go [hunk 12]
```diff
@@ -451,6 +451,22 @@ func (k *k8sSchedOps) UpgradePortworx(ociImage, ociTag string) error {
...
+func (k *k8sSchedOps) IsPXAppRunningOnNode(n node.Node) bool {
+	pxPods, err := k8s.Instance().GetPodsByNode(n.Name, PXNamespace)
+	if err != nil {
+		logrus.Errorf("Failed to get apps on node %s", n.Name)
+		return false
+	}
+	for _, pod := range pxPods.Items {
+		if strings.HasPrefix(pod.Name, "portworx") && pod.Status.Phase != "Running" {
```
Comment (harsh-px) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: use IsPodReady() api from https://github.com/portworx/sched-ops/k8s

Comment (harsh-px) [inline drivers/volume/portworx/schedops/k8s-schedops.go]: Instead of checking pod.name check pod labels. PX pods have label `name=portworx`.

Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Fixed.

Comment (Rohit-PX) [reply drivers/volume/portworx/schedops/k8s-schedops.go]: Fixed.

**Comment by Rohit-PX** [comment]
After changes requested by @harsh-px 👍 
http://70.0.0.41:8080/view/torpedo/view/all/job/torpedo-dev/73/console

**Pull request merged**